### PR TITLE
Added flag to check unique urls only once

### PR DESCRIPTION
**Pull request by ntotten**

### CODE REVIEW

File: lib/internal/defaultOptions.js [hunk 1]
```diff
@@ -16,7 +16,8 @@ var defaultOptions =
 	maxSocketsPerHost: 1,
 	rateLimit: 0,
 	tags: require("./tags"),
-	userAgent: userAgent(pkg.name, pkg.version)
+	userAgent: userAgent(pkg.name, pkg.version),
+	checkUniqueUrlOnce: false
```
Comment (stevenvachon) [inline lib/internal/defaultOptions.js]: Alphabetical order, please.

File: lib/public/UrlChecker.js [hunk 2]
```diff
@@ -63,24 +63,40 @@ UrlChecker.prototype.dequeue = function(id)
...
+		if (this.options.checkUniqueUrlOnce===true) {
+			if (this.urlCache.indexOf(url) > -1) {
+				return false;
+			} else {
+				this.urlCache.push(url);
+				return true;
+			}
+		}
+		return true;
+	}).bind(this);
```
Comment (stevenvachon) [inline lib/public/UrlChecker.js]: This deserves its own method.

Comment (stevenvachon) [inline lib/public/UrlChecker.js]: There should also be a `clearCache()` method in this class and pointed to from `HtmlUrlChecker` and `HtmlChecker`

File: lib/public/UrlChecker.js [hunk 3]
```diff
@@ -9,13 +9,13 @@ var RequestQueue = require("limited-request-queue");
...
 
 
-
 function UrlChecker(options, handlers)
 {
 	var thisObj = this;
 	
 	this.handlers = handlers || {};
 	this.options = parseOptions(options);
+	this.urlCache = [];
```
Comment (stevenvachon) [inline lib/public/UrlChecker.js]: `this.cache` would look nicer, I think.

File: lib/public/UrlChecker.js [hunk 4]
```diff
@@ -63,24 +63,40 @@ UrlChecker.prototype.dequeue = function(id)
...
-			url: url.url.parsed,
-			data: { customData:customData, linkObj:url }
-		});
+		if (testUniqueUrl(url.url.parsed.href)) {
+			return this.linkQueue.enqueue(
+			{
+				url: url.url.parsed,
+				data: { customData:customData, linkObj:url }
+			});
+		}
```
Comment (stevenvachon) [inline lib/public/UrlChecker.js]: `=== true`

File: lib/public/UrlChecker.js [hunk 5]
```diff
@@ -63,24 +63,40 @@ UrlChecker.prototype.dequeue = function(id)
...
 	// Documented use: enqueue(url, baseUrl)
 	// or erroneous and let linkQueue sort it out
 	else
 	{
-		return this.linkQueue.enqueue(
-		{
-			url: urllib.resolve(baseUrl || "", url),	// URL must be absolute
-			data: { orgUrl:url, baseUrl:baseUrl, customData:customData }
-		});
+		if (testUniqueUrl(url)) {
```
Comment (stevenvachon) [inline lib/public/UrlChecker.js]: `=== true`

File: test/4.public.UrlChecker.js [hunk 6]
```diff
@@ -199,5 +199,32 @@ describe("PUBLIC -- UrlChecker", function()
...
+
+			var instance = new UrlChecker( options,
+			{
+				link: function(result, customData)
+				{
+					results.push(result);
+				},
+				end: function()
+				{
+					expect(results).to.have.length(2);
```
Comment (stevenvachon) [inline test/4.public.UrlChecker.js]: Shouldn't it return the cached result? If a broken link exists in multiple places, all should be reported.

File: test/4.public.UrlChecker.js [hunk 7]
```diff
@@ -199,5 +199,32 @@ describe("PUBLIC -- UrlChecker", function()
...
+					expect(results[0].url.original).to.equal( conn.absoluteUrl+"/fixtures/index.html" );
+					expect(results[1].url.original).to.equal( conn.absoluteUrl+"/fixtures/link-real.html" );
+					done();
+				}
+			});
+			
+			instance.enqueue( conn.absoluteUrl+"/fixtures/index.html",     null, {index:0} );
+			instance.enqueue( conn.absoluteUrl+"/fixtures/index.html", 		 null, {index:1} );
```
Comment (stevenvachon) [inline test/4.public.UrlChecker.js]: white space issue

File: README.md [hunk 8]
```diff
@@ -194,6 +194,11 @@ Type: `Number`
 Default value: `0`  
 The number of milliseconds to wait before each request.
 
+### options.checkUniqueUrlOnce
```
Comment (stevenvachon) [inline README.md]: Perhaps `cacheResponses` would be a better name?

**Comment by stevenvachon** [comment]
@ntotten are you still interested in completing this?

**Comment by katerberg** [comment]
I've made the requested changes since this would be helpful for me on #8.

The only thing that I didn't agree with @stevenvachon on was that the request should be returned multiple times if cachedResponses is set to true. If the user specifies that they only want the request made once, I think it would be confusing to have a response reported multiple times. This is especially true since they are likely making this request with the understanding that the decreased verbosity means that some requests might be made multiple times.

Happy to discuss on #8.

**Pull request merged**